### PR TITLE
ApiKey owner (org) scope UI work

### PR DESCRIPTION
**Pull request by chenriksson**
Issue #4872.

Work remaining for API key owner (org) scopes:
- Auditing (#4958)
- Remove deprecated Scope.cs ctor (will do in separate PR, to minimize this PR)
- Integrate with PermissionsService

### CODE REVIEW

**Comment by chenriksson** [comment]
- Create key has same UI if user is not member of any organization:

image

- Existing v2 key w/o owner scope:

image

- Existing v2 key with owner scope:

image

- Creation of new v2 key, will always have owner scope:

image

- Other Notes:
  - Push new action disabled for organization collaborators
  - Packages displayed depends on the selected owner

File: src/NuGetGallery/Controllers/UsersController.cs [hunk 1]
```diff
@@ -558,14 +568,24 @@ public virtual ActionResult PasswordChanged()
         [Authorize]
         [HttpPost]
         [ValidateAntiForgeryToken]
-        public virtual async Task<JsonResult> GenerateApiKey(string description, string[] scopes = null, string[] subjects = null, int? expirationInDays = null)
+        public virtual async Task<JsonResult> GenerateApiKey(string description, string owner, string[] scopes = null, string[] subjects = null, int? expirationInDays = null)
```
Comment (scottbommarito) [inline src/NuGetGallery/Controllers/UsersController.cs]: I think we should change `scopes` to be `actions` here and elsewhere. It's incredibly confusing that a `Scope` is an object that contains an owner, a set of actions, and a list of subjects AND a "scope" is an action that the API key allows on the package. Now we're also referring to the owner as a "scope" too.

File: src/NuGetGallery/Scripts/gallery/page-api-keys.js [hunk 2]
```diff
@@ -106,13 +101,33 @@
...
             this.PendingDescription = ko.observable();
+
+            // Package owner selection
+            this.PackageOwner = ko.observable();
+            this.PackageOwnerName = ko.pureComputed(function () {
+                return self.PackageOwner() && self.PackageOwner().Owner;
+            }, this);
+
+            // Do not default to push new scope if disabled (organization collaborator)
+            this.PushNewEnabled = ko.pureComputed(function () {
```
Comment (scottbommarito) [inline src/NuGetGallery/Scripts/gallery/page-api-keys.js]: You will probably want to do this for all three possible actions (push new, push version, and unlist) because the `PermissionsService` allows for it.

File: src/NuGetGallery/Views/Users/ApiKeys.cshtml [hunk 3]
```diff
@@ -174,6 +174,10 @@
                         <!-- /ko -->
                     </li>
                 </ul>
+                <!-- ko if: Owner -->
+                <b>Package owner:</b> <span data-bind="text: Owner"></span>
```
Comment (scottbommarito) [inline src/NuGetGallery/Views/Users/ApiKeys.cshtml]: any thoughts on showing the current user if there is no owner scope?

File: src/NuGetGallery/Controllers/UsersController.cs [hunk 4]
```diff
@@ -558,14 +568,24 @@ public virtual ActionResult PasswordChanged()
...
         {
             if (string.IsNullOrWhiteSpace(description))
             {
                 Response.StatusCode = (int)HttpStatusCode.BadRequest;
                 return Json(Strings.ApiKeyDescriptionRequired);
             }
 
+            // Get the owner scope
+            User scopeOwner = string.IsNullOrEmpty(owner)
```
Comment (scottbommarito) [inline src/NuGetGallery/Controllers/UsersController.cs]: We need to check that the current user actually has the permissions to create an API key with this user as their owner scope. From what I can see here, someone could create an API key with any user as the owner in the scope.

On a similar note, we should check that the user has permissions to create an API key with these scopes (actions) on the owner (e.g. organization collaboration shouldn't have push new).

File: src/NuGetGallery/Views/Users/ApiKeys.cshtml [hunk 5]
```diff
@@ -293,6 +297,23 @@
                 </div>
             }
         </div>
+
+        @if (Model.PackageOwners.Count > 1)
```
Comment (scottbommarito) [inline src/NuGetGallery/Views/Users/ApiKeys.cshtml]: If `Model.PackageOwners.Count == 1` you may want to pass the single option as a hidden item on the form.

It's not needed because you have the `GetCurrentUser` call in `GenerateApiKey`, but in case some code changes make it possible for the only package owner to be NOT the current user, this is more consistent.

File: src/NuGetGallery/ViewModels/ScopeViewModel.cs [hunk 6]
```diff
@@ -5,11 +5,13 @@ namespace NuGetGallery
...
     public class ScopeViewModel
     {
+        public string Owner { get; set; }
         public string Subject { get; set; }
         public string AllowedAction { get; set; }
 
-        public ScopeViewModel(string subject, string allowedAction)
+        public ScopeViewModel(string owner, string subject, string allowedAction)
         {
+            Owner = owner ?? "*";
```
Comment (scottbommarito) [inline src/NuGetGallery/ViewModels/ScopeViewModel.cs]: `"*"` -> `NuGetPackagePattern.AllInclusivePattern`

File: src/NuGetGallery/Controllers/UsersController.cs (hunk 1 above)
Comment (chenriksson) [reply src/NuGetGallery/Controllers/UsersController.cs]: I agree. The term "scope" is used interchangeably with scope action throughout the code.

I'm ok with fixing, but prefer to do it in a separate PR as I don't want to inflate this one.

File: src/NuGetGallery/Scripts/gallery/page-api-keys.js (hunk 2 above)
Comment (chenriksson) [reply src/NuGetGallery/Scripts/gallery/page-api-keys.js]: Push version and unlist are always available, even to organization collaborators. I only need to determine when to enable/disable push new.

File: src/NuGetGallery/Views/Users/ApiKeys.cshtml (hunk 3 above)
Comment (chenriksson) [reply src/NuGetGallery/Views/Users/ApiKeys.cshtml]: We decided that no owner scope means that it can match any, not just the current user. That's why I chose to show "*"... but am open to a different suggestion.

File: src/NuGetGallery/ViewModels/ScopeViewModel.cs (hunk 6 above)
Comment (chenriksson) [reply src/NuGetGallery/ViewModels/ScopeViewModel.cs]: It's not a package pattern, but will add AllInclusivePattern const to the view model.

File: src/NuGetGallery/Controllers/UsersController.cs (hunk 4 above)
Comment (joelverhagen) [reply src/NuGetGallery/Controllers/UsersController.cs]: Yes, client side validation is a "hint" to the user. Server side validation is the real enforcement.

File: src/NuGetGallery/Scripts/gallery/page-api-keys.js [hunk 7]
```diff
@@ -106,13 +101,33 @@
...
+                return self.PackageOwner() && self.PackageOwner().CanPushNew;
+            }, this);
+            this.PushNewEnabled.subscribe(function (newValue) {
+                var defaultPushId = (newValue)
```
Comment (joelverhagen) [inline src/NuGetGallery/Scripts/gallery/page-api-keys.js]: nit: why parens around `newValue`?

File: src/NuGetGallery/Controllers/UsersController.cs [hunk 8]
```diff
@@ -558,14 +568,24 @@ public virtual ActionResult PasswordChanged()
...
+                ? GetCurrentUser()
+                : _userService.FindByUsername(owner);
+            if (scopeOwner == null)
```
Comment (joelverhagen) [inline src/NuGetGallery/Controllers/UsersController.cs]: Just making sure -- this code is saying all future API keys must be scoped to a single owner?

**Comment by joelverhagen** [comment]
We need a final sign-off from all feature owners and @anangaur that:

1. What do API keys without an owner scope do when pushing a new ID and the user is part of at least one org. _This is the ambiguous case..._
1. What should the API key UI show for the scopes/owners/actions for these old owner-less API keys?

I keep forgetting our conclusion on these discussions 😊.

File: src/NuGetGallery/Scripts/gallery/page-api-keys.js (hunk 2 above)
Comment (scottbommarito) [reply src/NuGetGallery/Scripts/gallery/page-api-keys.js]: always available NOW

in the future they may not be, so we use `PermissionsService` to centralize this logic

File: src/NuGetGallery/Views/Users/ApiKeys.cshtml (hunk 3 above)
Comment (scottbommarito) [reply src/NuGetGallery/Views/Users/ApiKeys.cshtml]: Oh my bad, I will need to update my PR to understand this logic

File: src/NuGetGallery/Views/Users/ApiKeys.cshtml (hunk 5 above)
Comment (scottbommarito) [reply src/NuGetGallery/Views/Users/ApiKeys.cshtml]: https://github.com/NuGet/NuGetGallery/pull/4963/files#diff-5058b8e804fd7b6e9e60a29de7f6fff6R60 is how I did it for upload/edit

**Comment by scottbommarito** [comment]
@joelverhagen 
question 1 is handled in https://github.com/NuGet/NuGetGallery/pull/4963 (we should ask it there)
not sure about question 2

**Comment by anangaur** [comment]
> What do API keys without an owner scope do when pushing a new ID and the user is part of at least one org. This is the ambiguous case...
What should the API key UI show for the scopes/owners/actions for these old owner-less API keys?

Both cases are for existing API keys and they should get associated to the account owner's account, UI or otherwise. The spec has this discussed here: https://github.com/NuGet/Home/wiki/Organizations-on-NuGet.org#what-happens-to-authors-existing-api-keys

File: src/NuGetGallery/Scripts/gallery/page-api-keys.js [hunk 9]
```diff
@@ -139,7 +154,7 @@
                     return id;
                 }, self);
             }
-
+
```
Comment (scottbommarito) [inline src/NuGetGallery/Scripts/gallery/page-api-keys.js]: nit: unnecessary whitespace

File: src/NuGetGallery/Views/Users/ApiKeys.cshtml (hunk 3 above)
Comment (scottbommarito) [reply src/NuGetGallery/Views/Users/ApiKeys.cshtml]: according to @anangaur, we decided that no owner scope means only match current user

personally, I prefer this because it forces users to use organizations correctly

File: src/NuGetGallery/Controllers/UsersController.cs [hunk 10]
```diff
@@ -558,14 +568,24 @@ public virtual ActionResult PasswordChanged()
...
+            {
+                Response.StatusCode = (int)HttpStatusCode.NotFound;
```
[the remark this answers is not in the document]
Comment (chenriksson) [reply src/NuGetGallery/Controllers/UsersController.cs]: Updated

File: src/NuGetGallery/Scripts/gallery/page-api-keys.js (hunk 7 above)
Comment (chenriksson) [reply src/NuGetGallery/Scripts/gallery/page-api-keys.js]: Updated

File: src/NuGetGallery/ViewModels/ScopeViewModel.cs (hunk 6 above)
Comment (chenriksson) [reply src/NuGetGallery/ViewModels/ScopeViewModel.cs]: Removed per @anangaur feedback that legacy keys match only current user.

File: src/NuGetGallery/Controllers/UsersController.cs (hunk 8 above)
Comment (chenriksson) [reply src/NuGetGallery/Controllers/UsersController.cs]: Yes, API keys will be scoped to a single owner.

File: src/NuGetGallery/Controllers/UsersController.cs [hunk 11]
```diff
@@ -105,23 +105,33 @@ public virtual ActionResult ApiKeys()
...
                 .ToList();
 
-            // Get package IDs
-            var packageIds = _packageService
-                .FindPackageRegistrationsByOwner(user)
-                .Select(p => p.Id)
-                .OrderBy(i => i)
+            // Get package owners (user's self or organizations)
+            var owners = user.Organizations
+                .Select(o => CreateApiKeyOwnerViewModel(o.Organization, canPushNew: o.IsAdmin))
```
[the remark this answers is not in the document]
Comment (chenriksson) [reply src/NuGetGallery/Controllers/UsersController.cs]: Will integrate w/ PermissionsService in a follow-up PR.

File: src/NuGetGallery/Controllers/UsersController.cs (hunk 4 above)
Comment (chenriksson) [reply src/NuGetGallery/Controllers/UsersController.cs]: Adding server side validation, but will integrate w/ PermissionsService in follow-up PR since it's under flux right now.

File: src/NuGetGallery/Views/Users/ApiKeys.cshtml (hunk 5 above)
Comment (chenriksson) [reply src/NuGetGallery/Views/Users/ApiKeys.cshtml]: Modified to still populate PackageOwner observable when single owner (self), but will hide the drop-down.

File: src/NuGetGallery/Scripts/gallery/page-api-keys.js (hunk 9 above)
Comment (chenriksson) [reply src/NuGetGallery/Scripts/gallery/page-api-keys.js]: reverted

**Comment by chenriksson** [comment]
@scottbommarito @joelverhagen Addressed feedback. I will integrate it with PermissionsService in a separate PR, to avoid taking dependency for now on #4963.

There's still a test I'd like to add for server-side validation, but I'm currently blocked there by a VS bug with test discovery.

File: src/NuGetGallery/Authentication/AuthenticationService.cs [hunk 12]
```diff
@@ -524,7 +527,11 @@ public virtual CredentialViewModel DescribeCredential(Credential credential)
                 // Set the description as the value for legacy API keys
                 Description = credential.Description, 
                 Value = kind == CredentialKind.Token && credential.Description == null ? credential.Value : null,
-                Scopes = credential.Scopes.Select(s => new ScopeViewModel(s.Subject, NuGetScopes.Describe(s.AllowedAction))).ToList(),
+                Scopes = credential.Scopes.Select(s => new ScopeViewModel(
+                        scopeOwner.Username,
```
Comment (scottbommarito) [inline src/NuGetGallery/Authentication/AuthenticationService.cs]: You shouldn't cache the value and use it for each scope in the list when you're enumerating through each scope anyway.

```
s.Owner?.Username ?? credential.User
```

File: src/NuGetGallery/Extensions/ScopeExtensions.cs [hunk 13]
```diff
@@ -70,6 +71,19 @@ public static bool HasOwnerScope(this Scope scope)
 
             return scope.OwnerKey.HasValue;
         }
+
+        public static User GetOwnerScope(this IEnumerable<Scope> scopes)
```
Comment (scottbommarito) [inline src/NuGetGallery/Extensions/ScopeExtensions.cs]: I would suggest removing this function and using `scopes.SingleOrDefault(s => s.Owner)` instead in all the places that you use it for the following reasons.

- it can be reduced into the one-liner above
- it will need to be removed when we allow multiple owners per scope
- its existence makes it very easy to not treat scopes appropriately (e.g. by assuming that the first owner in the scope applies to ALL scopes)

File: src/NuGetGallery/Strings.resx [hunk 14]
```diff
@@ -618,4 +618,7 @@ For more information, please contact '{2}'.</value>
   <data name="AccountDelete_Success" xml:space="preserve">
     <value>The account:{0} was deleted succesfully.</value>
   </data>
+  <data name="ApiKeyOwnerRequired" xml:space="preserve">
+    <value>Can't generate an API key without a package owner.</value>
```
Comment (scottbommarito) [inline src/NuGetGallery/Strings.resx]: `You must specify a package owner to generate an API key.` might be better

File: src/NuGetGallery/Extensions/ScopeExtensions.cs (hunk 13 above)
Comment (chenriksson) [reply src/NuGetGallery/Extensions/ScopeExtensions.cs]: IMHO, if we eventually allow multiple scope owners, having this in a central location would actually help identify callers that need updating.

File: src/NuGetGallery/Strings.resx (hunk 14 above)
Comment (chenriksson) [reply src/NuGetGallery/Strings.resx]: Was consistent with ApiKeyDescriptionRequired, but I have no strong preference. Updated.

File: src/NuGetGallery/ViewModels/ApiKeyViewModel.cs [hunk 15]
```diff
@@ -20,6 +20,13 @@ public ApiKeyViewModel(CredentialViewModel cred)
                 throw new ArgumentNullException(nameof(cred));
             }
 
+            // Currently ApiKeys.cshtml has single Owner per ApiKey restriction.
+            var owner = cred
```
Comment (joelverhagen) [inline src/NuGetGallery/ViewModels/ApiKeyViewModel.cs]: Can you use `GetOwnerScope`  extension method here?

Comment (chenriksson) [reply src/NuGetGallery/ViewModels/ApiKeyViewModel.cs]: No... code looks the same, but this is querying on view models not entities.
